private_gpt/server/chat/chat_service.py

```diff
@@ -58,63 +57,69 @@ def __init__(
-        return ContextChatEngine.from_defaults(
-            retriever=vector_index_retriever,
-            service_context=self.service_context,
-            node_postprocessors=[
-                MetadataReplacementPostProcessor(target_metadata_key="window"),
-            ],
-        )
-
-    def stream_chat(
         self,
-        messages: list[ChatMessage],
+        system_prompt: str | None = None,
         use_context: bool = False,
         context_filter: ContextFilter | None = None,
-    ) -> CompletionGen:
+    ) -> BaseChatEngine:
         if use_context:
-            last_message = messages[-1].content
-            chat_engine = self._chat_engine(context_filter=context_filter)
-            streaming_response = chat_engine.stream_chat(
-                message=last_message if last_message is not None else "",
-                chat_history=messages[:-1],
+            vector_index_retriever = self.vector_store_component.get_retriever(
+                index=self.index, context_filter=context_filter
             )
-            sources = [
-                Chunk.from_node(node) for node in streaming_response.source_nodes
-            ]
-            completion_gen = CompletionGen(
-                response=streaming_response.response_gen, sources=sources
+            return ContextChatEngine.from_defaults(
+                system_prompt=system_prompt,
+                retriever=vector_index_retriever,
+                service_context=self.service_context,
+                node_postprocessors=[
+                    MetadataReplacementPostProcessor(target_metadata_key="window"),
+                ],
             )
         else:
-            stream = self.llm_service.llm.stream_chat(messages)
-            completion_gen = CompletionGen(
-                response=stream_chat_response_to_tokens(stream)
+            return SimpleChatEngine.from_defaults(
+                system_prompt=system_prompt,
+                service_context=self.service_context,
             )
+
+    def stream_chat(
+        self,
+        messages: list[ChatMessage],
+        system_prompt: str | None = None,
+        use_context: bool = False,
+        context_filter: ContextFilter | None = None,
+    ) -> CompletionGen:
+        last_message = messages[-1].content
+        chat_engine = self._chat_engine(
+            system_prompt=system_prompt,
+            use_context=use_context,
+            context_filter=context_filter,
+        )
+        streaming_response = chat_engine.stream_chat(
+            message=last_message if last_message is not None else "",
+            chat_history=messages[:-1],
+        )
+        sources = [Chunk.from_node(node) for node in streaming_response.source_nodes]
+        completion_gen = CompletionGen(
+            response=streaming_response.response_gen, sources=sources
+        )
         return completion_gen
 
     def chat(
         self,
         messages: list[ChatMessage],
+        system_prompt: str | None = None,
```

Comment:
`system_prompt` can be passed in `ChatMessage`, as the first message.

Doing this would reduce the number of moving pieces and the number of parameters in function.

This also contains the "state" exclusively in the list of messages